W-10804256-dw-extension-plugin-kt (#114)
mulesoft/docs-dataweave: modules/ROOT/pages/dataweave-extension-plugin.adoc

//image::dataweave-extension-testmapping.png[]
image::animations/mappingTest.gif[Test Mapping Steps]
// ![](animations/mappingTest.gif)

// == Deploy and Share DataWeave Libraries
// This section does not exist in the DW Extension plugin documentation but I consider that we should add it, even though the deployment is through the maven plugin.

== Use Playground Mode
If you want to try out your mapping without creating a DataWeave project you can open your `.dwl` mapping file directly on Visual Studio Code or create a new one and use scenarios and sample data.

== Use DataWeave Dependencies View

The *DATAWEAVE DEPENDENCIES* view shows all the dependencies resolved for a DataWeave project. The view enables you to navigate through all of your project's dependencies and open DataWeave files inside those dependencies.

//image::dataweave-extension-weavedependencies.png[]
image::animations/weaveDependencie.gif[Weave Dependencies view]
// ![](animations/weaveDependencies.gif)

== DataWeave Extension Plugin Language Edition Features

Check the following edition features that you can use in Visual Studio Code:

=== Completion

> The completion feature enables you to auto-complete content for any visible functions, variables, or types. The feature also suggests fields based on the type inference.

Reviewer: ```suggestion
The completion feature enables you to autocomplete content for any visible functions, variables, or types. The feature also suggests fields based on the type inference.
```

Reviewer: Do we need to explain what type inference means, or provide a "for instance"?

//image::dataweave-extension-completion.png[]
image::animations/completion.gif[Completion Edition Feature]
// ![](animations/completion.gif)

=== Navigation

> The navigation feature enables you to navigate to any local definition or from any imported library.

Reviewer: When/why would you want or need to use this?

//image::dataweave-extension-navigation.png[]
image::animations/navegation.gif[Navigation Edition Feature]
// ![](animations/navegation.gif)

=== Code Editing Features

The following features help you to easily code in Visual Studio Code:

* Find all references
* Refactors (local and cross files)
* Quick fixes:
** Auto import when a function is present in a module
** Create a function, variable or type when is not found
* Outline
* Show parameters information
* Hover information with documentation

//image::dataweave-extension-autoimport.png[]
image::animations/auto_import.gif[Auto Import Code Editing Feature]

=== Code Inspections

> The DataWeave Extension plugin tries to detect idioms that you express and suggests best practices, for example:

Reviewer: ```suggestion
The DataWeave Extension plugin tries to detect and suggest replacements for idioms in expressions, as shown in the following examples:
```

= Installing and Using the DataWeave Extension Plugin

> Install and use the DataWeave Extension plugin in Visual Studio Code to develop, preview, test, and deploy custom mappings and modules into libraries in a DataWeave project. The extension plugin also enables you to perform transformations similar to those you perform in the https://developer.mulesoft.com/learn/dataweave/[DataWeave Playground].

Reviewer 2: depending on finalized name. check with Esteben maybe

== Before You Begin

Before installing the DataWeave Extension plugin, you must:

> * Install JDK (Java 8 minimum)

Reviewer 2: Or actually, it's more like "DataWeave 2.0 Extension to Visual Studio Code" 

Perhaps we could shorten after first mention to "DataWeave extension"?

> * Install JDK (Java 8 minimum)

Reviewer 3: I think we should recommend AdoptOpenJDK, since it's the one that we have on other products, and the one we are using on our build.

> = Installing and Using the DataWeave Extension Plugin

Reviewer 2: Is "DataWeave Extension Plugin" the name. Extension = Plugin 
VS Code uses "extension". 

Maybe "DataWeave Extension to Visual Studio Code"

Install and use the DataWeave 2.0 extension in Visual Studio Code to develop, preview, test, and deploy custom mappings and modules into libraries in a DataWeave project. The extension also enables you to perform transformations similar to those you perform in the https://developer.mulesoft.com/learn/dataweave/[DataWeave Playground].

== Before You Begin

Before installing the DataWeave 2.0 extension, you must:

* Install JDK (Java 8 minimum)

> * Install Apache Maven

Reviewer 2: Any particular version? Perhaps the latest or the one recommended by VS Code?
Asking bc I had Maven installed and then was asked by Visual Studio Code if I wanted
to install the recommended version of it, which I did. The version was 17.0.2

Author: @martincousido

Reviewer 3: We should require at least 3.6.3 which is the embedded version we use for internally to resolve some dependencies.

[start=5]
. Follow prompts to complete an initial configuration for your project:
+
.. *Organization ID* (Group ID)
.. *Artifact ID*
.. *Version*
.. *Project name*
[start=5]
. Choose a directory folder for your new project.
. Select *OK*.
. When prompted, select *Yes* or *No* whether you want to open the project in a new Visual Studio Code window.

The following image shows a new DataWeave project in Visual Studio Code:

image::dataweave-extension-new-project-sample.png[DataWeave project example window]

//image::animations/new_project.gif[New project creation steps]
//![](animations/new_project.gif)

When creating a DataWeave project, the DataWeave Extension plugin requires all dependencies to be loaded and indexed before it can search the libraries and make them available in your project. Visual Studio Code provides a notification bar at the bottom of the screen editor that shows when the dependencies are loading and when the indexing is complete.

image::dataweave-extension-dependencies.png[Dependencies loading and indexing]
//image::animations/loading.gif[Dependencies loading and indexing]
// ![](animations/loading.gif)

== Create a DataWeave Module

> DataWeave modules are `.dwl` files that define functions, variables, types, and namespaces. These files can't contain output directives, body expressions, or `---` separators between header and body sections. A DataWeave module enables you to share all the content defined in the `.dwl` file for reuse.

Reviewer 2: Do we allow contractions like "can't"?

To create a new DataWeave module, follow these steps:

> . In Visual Studio Code, launch *Command Palette* by pressing `Ctrl + Shift + P` (Windows, Linux) or `⇧⌘P` (macOS).

Reviewer 2: same Q here re whether you can open the palette with View -> Command Palette on all variants.

. In the search bar, type `dataweave`.
. Select *DataWeave: Create New Module*.
+

image::dataweave-extension-createmodule.png[Command Palette with DataWeave: Create new module option selected]
//image::animations/new_mapping.gif[Creating new Module using DataWeave Extension Plugin]
// ![](animations/new_mapping.gif)

== Test a DataWeave Module

> There are two ways to test your module: either by running a unit test or by creating a test mapping located on your `src/test/dw` folder as it should not be exposed to consumers of your module.

Reviewer 2: ```suggestion
There are two ways to test your module: 

* By running a unit test
* By creating a test mapping located in your `src/test/dw` folder to avoid exposing the test to to consumers of your module
```

> The tests use functions of the modules `dw::test::Tests` and `dw::test::Asserts` provided by the artifact `org.mule.weave.wtf` of the

Reviewer 2: Do we have descriptions of the functions in those modules? Will they be referenced in the testing framework doc maybe?

Author: The DW Testing framework doc does not have descriptions for  `dw::test::Tests` and `dw::test::Asserts`. Could you provide descriptions @martincousido? Thank you!

Reviewer 4: Mmmm, I'm not sure if we need to provide the description in the doc. But the functions of those modules can be hover in the tests files .dwl and the user can see the description of what the different functions do. 
For example, `describedBy` "Defines a new test suite with the list of test cases."

Reviewer 4: ```suggestion
The tests use functions of the modules `dw::test::Tests` and `dw::test::Asserts` provided by the artifact `org.mule.weave:data-weave-testing-framework` of the
```
Renamed from wtf to data-weave-testing-framework. Not sure if `:` is used normally to separate groupId from artifactId.

//xref:dataweave-testingframework.adoc[DataWeave Testing Framework].

To generate a unit test for a DataWeave module, follow these steps:

. In Visual Studio Code, open your DataWeave project.
. Navigate to the `src/main/dw` folder.
. Select your module `.dwl` file.
. In your module `.dwl` file, click the *Add Unit Test* CodeLens button located above your function.
+
image::dataweave-extension-testmodule-1.png[MyModule.dwl file showing the function and the Add Unit Test button]

//image::animations/moduleTest.gif[Test Module Steps]
// ![](animations/moduleTest.gif)

This behavior generates:

> * A new `.dwl` test file for the module under the `src/test/dw` folder. The test file defines a test case for the function that you call with specific arguments and make assertions over the output.

Reviewer 2: "test case for the function that you call ... and make assertions over the output" 
This doesn't sound right.  

maybe  "... and use to make assertions over..." if referring to what the test case ?

* Either a new test suite for the function, consisting of your new test case, or a new test case to an existing test suite.

> // > Dev team please define/explain better test file vs test case vs test suite

Author: @martincousido Could you clarify Duke's question above and also provide definitions for the following terms?:

- test file
- test case
- test suite

Reviewer 3: Test file -> The .dwl file where test cases and test suites are described.
Test suite -> This is a way of grouping tests with a name that better describes the intention of some related cases.
Test case -> A single test, with the evaluation and assertion of the result.

== Create a DataWeave Mapping

A DataWeave mapping is a `.dwl` file that enables you to map and connect to anything, regardless of data type or structure. As when using a DataWeave module, you can use a mapping to define functions, variables, types, and namespaces, but you can additionally specify a body section after the separator `---`.

Unlike a DataWeave module file, a DataWeave mapping file is an executable unit that enables you to transform zero or more inputs into a single output. Using the DataWeave 2.0 extension, you can develop these files to create reusable assets that you can deploy and share with others.

To create a new DataWeave mapping, follow these steps:

. In Visual Studio Code, launch *Command Palette* by pressing `Ctrl + Shift + P` (Windows, Linux) or `⇧⌘P` (macOS).
. In the search bar, type `dataweave`.
. Select *DataWeave: Create New Mapping*.

image::dataweave-extension-createmapping.png[Command Palette with DataWeave: Create new mapping selected]
//image::animations/new_mapping.gif[Creating new mapping using DataWeave Extension Plugin]
// ![](animations/new_mapping.gif)

[[sample-data]]
=== Define Sample Data for DataWeave Mappings

> Generate multiple sample data inputs for your mapping by clicking the *Define Sample Data* CodeLens button at the top of your mapping `.dwl` file. Sample data is a key part of DataWeave project development, because it is used to run your mappings and also for suggestions.

Reviewer 2: Do people know what CodeLens buttons are? I had to look it up. But maybe that's just me. (ha,ha)

Author: I also had to look it up in VS Code documentation. I think we can remove the wording "CodeLens" @dukesphere 
The screenshots highlight the buttons.

Reviewer 3: It is a concept introduced by VSCode for links inside the code editor that trigger different actions.
People not coming from vscode environments might have trouble understanding what a CodeLens is. Should we add a short description somewhere?

. Navigate to the `src/main/dw` folder.
. Select your mapping `.dwl` file.
. Click the *Run Mapping* button located at the top of your `.dwl` file.
+
image::dataweave-extension-runmapping.png[MyMapping.dwl and Run Mapping button]

//image::animations/debug.gif[Run and debug a mapping]
// ![](animations/debug.gif)

== Test a DataWeave Mapping

To test a DataWeave mapping, click the *Dataweave: Create Mapping Test* button located at the top of your `.dwl` file to create a new test scenario from a snapshot of your input and your current output.
You can also use this newly created scenario as a normal scenario for your preview or autocompletion.

To test a DataWeave mapping, follow these steps:

. In Visual Studio Code, open your DataWeave project.
. Navigate to the `src/main/dw` folder.
. Click the *DataWeave: Create Mapping Test* button (microscope icon) located at the top of your mapping `.dwl` file.
+
image::dataweave-extension-testmapping.png[MyMapping.dwl file and the DataWeave Create Mapping Test button]
[start=4]
. Specify a name for the mapping test.

//image::animations/mappingTest.gif[Test Mapping Steps]
// ![](animations/mappingTest.gif)

> // == Deploy and Share DataWeave Libraries

Reviewer 2: I agree that all doc that relates to the extension should be part of the extension doc.

image::dataweave-extension-new-project-sample.png[DataWeave project example window]

When creating a DataWeave project, the DataWeave 2.0 extension requires all dependencies to be loaded and indexed before it can search the libraries and make them available in your project. Visual Studio Code provides a notification bar at the bottom of the screen editor that shows when the dependencies are loading and when the indexing is complete.

image::dataweave-extension-dependencies.png[Dependencies loading and indexing]

== Create a DataWeave Module

DataWeave modules are `.dwl` files that define functions, variables, types, and namespaces. These files cannot contain output directives, body expressions, or `---` separators between header and body sections. A DataWeave module enables you to share all the content defined in the `.dwl` file for reuse.

To create a new DataWeave module:

. In Visual Studio Code, click *View > Command Palette*.
. In the search bar, type `dataweave`.
. Select *DataWeave: Create New Module*.
+

image::dataweave-extension-createmodule.png[Command Palette with DataWeave: Create new module option selected]

> 

Reviewer 4: Maybe we can show one example for developing a function of a module (without using "sample data", which at this point it is not explained yet)

> 

Reviewer 4: <img width="819" alt="Screen Shot 2022-03-17 at 6 24 59 PM" src="https://user-images.githubusercontent.com/8797246/158897815-6f33ad6d-79d6-4e50-896b-feef4d76ba0d.png">

<img width="736" alt="Screen Shot 2022-03-17 at 6 26 09 PM" src="https://user-images.githubusercontent.com/8797246/158897857-c550fc96-a436-4ba2-88e2-2129436d4104.png">

│   │   └── resources
│   └── test
│       ├── dw
│       │   └── Test.dwl //Unit tests go here
│       │   └── MyMapping // Name of your mapping
│       │       └── default //Name of your scenario
│       │           └── inputs //Folder where all of your inputs live
│       │           │   └── payload.json //Input for your mapping (you can use it in your script, naming it to be the same as the name of the file without the file's extension)
│       │           │   └── vars //Folder that represents a name for your input variable in your script: in this case ,vars
│       │           │       └── test.json //Input for your mapping (you can use it in your script, calling it vars.test)
│       │           └── out.json // Expected output of your mapping running against your inputs (only present when testing)
```
=== Using the DataWeave Scenarios View

The *DATAWEAVE SCENARIOS* view shows the available scenarios for the `.dwl` file. In this view, you can add, remove, or set the active scenario that is used for your preview execution, or you can choose that the values be autocompleted.

image::dataweave-extension-dataweavescenario.png[DataWeave Scenarios view in Visual Studio Code]

> == Preview a DataWeave Mapping

Reviewer 4: Remove this section if it is added previously in the suggestion of "Develop using the Preview"

After you define sample data, get feedback on how your mapping is working by previewing the output. To do so, follow one of these options:

* Click the *DataWeave: Run Preview* button (arrow icon) at the top right corner of your mapping `.dwl` file window.
+
image::dataweave-extension-preview-1.png[DataWeave: Run Preview button]

* Right-click the editor and select *DataWeave: Enable AutoPreview* to run it on every file change.
+
image::dataweave-extension-preview-2.png[DataWeave: Enable AutoPreview option]

== Run and Debug a DataWeave Mapping

Run and debug a DataWeave mapping to be guided through all of the mapping elements, which you can then fix or otherwise update. You can also set breakpoints to stop evaluating expressions and use all the debugging capabilities that Visual Studio Code provides. Debugging works on mapping tests as well.

To run and debug a DataWeave mapping, follow these steps:

. In Visual Studio Code, open your DataWeave project.
. Navigate to the `src/main/dw` folder.
. Select your mapping `.dwl` file.
. Click the *Run Mapping* button located at the top of your `.dwl` file.
+

> image::dataweave-extension-runmapping.png[MyMapping.dwl and Run Mapping button]

Reviewer 4: Could we remove the input directive here?

== Test a DataWeave Mapping

To test a DataWeave mapping, click the *Dataweave: Create Mapping Test* button located at the top of your `.dwl` file to create a new test scenario from a snapshot of your input and your current output.
You can also use this newly created scenario as a normal scenario for your preview or autocompletion.

To test a DataWeave mapping, follow these steps:

. In Visual Studio Code, open your DataWeave project.
. Navigate to the `src/main/dw` folder.
. Click the *DataWeave: Create Mapping Test* button (microscope icon) located at the top of your mapping `.dwl` file.
+

> image::dataweave-extension-testmapping.png[MyMapping.dwl file and the DataWeave Create Mapping Test button]

Reviewer 4: Could we avoid the input directive in the image?
It would be nice to show the output of the preview here, so it is clear that we are saving the mapping scenario with the output of the Preview as a test.

[start=4]
. Specify a name for the mapping test.

== Using the DataWeave Dependencies View

The *DATAWEAVE DEPENDENCIES* view shows all the dependencies resolved for a DataWeave project. The view enables you to navigate through all of your project's dependencies and open DataWeave files inside those dependencies.

> image::dataweave-extension-dataweavedependencies.png[DataWeave Dependencies View]

Reviewer 4: Could be changed to show Core.dwl which is the normal use case for navigating to dependencies?

== Deploy DataWeave Libraries

Before deploying your DataWeave library to Anypoint Exchange, generate documentation for your mapping or module functions. When you deploy your DataWeave library to Exchange, the documentation is also published to the Exchange library page.

To generate documentation for the functions, follow these steps:

. In Visual Studio Code, open your DataWeave project.
. Navigate to the `src/main/dw` folder.
. Click the *Generate Weave Documentation* button located at the top of the function in the `.dwl` file.
+

> image::dataweave-extension-generatedocs-1.png[MyMapping.dwl file and the Generate Weave Documentation button]

Reviewer 4: This should be a module. Not sure if docs for functions inside mappings are rendered in exchange because those functions are not exposed.

The documentation template appears as a comment above the function:

> image::dataweave-extension-generatedocs-2.png[Auto-generated documentation located above the function]

Reviewer 4: This should be a module. Not sure if docs for functions inside mappings are rendered in exchange because those functions are not exposed.

+
For guidance, refer to the xref:dataweave-maven-plugin.adoc[DataWeave Maven Plugin] documentation.

== Consume DataWeave Libraries

After deploying your DataWeave library to Exchange, the library appears as an asset for your organization in the Exchange marketplace.

To consume the DataWeave library:

. Add the library’s group ID, artifact ID, version, and classifier to the `dependencies` element in your project’s `pom.xml` file.
+
You can copy the dependency snippet from Exchange.

. Add the Maven facade as a repository in the `repositories` section.

For more details, refer to the xref:exchange::to-publish-assets-maven.adoc#publish-and-consume-federated-assets[Publish and Consume Federated Assets] documentation.

== DataWeave 2.0 Extension Language Edition Features

The Language Edition features of the DataWeave 2.0 extension include completion, navigation, code editing, and code inspection.

=== Completion

The completion feature enables you to autocomplete content for any visible functions, variables, or types. The feature also suggests fields based on the type inference.

> image::dataweave-extension-completion.png[Completion Feature sample]

Reviewer 4: Can we remove the input directive?
And also remove the parameters? because that is autocompleted too when you call to `filter`

=== Navigation

> You to navigate to any local definition or from any imported library.

Reviewer 4: ```suggestion
You can navigate to any local definition or from any imported library.
```

=== Code Editing Features

The following features help you to easily code in Visual Studio Code:

* Find all references
* Refactors (local and cross files)
* Quick fixes:
** Auto import when a function is present in a module
** Create a function, variable, or type when one is not found
* Outline
* Show parameters information
* Hover information with documentation

image::dataweave-extension-autoimport.png[Auto Import Feature sample]

=== Code Inspections

The DataWeave 2.0 extension tries to detect and suggest replacements for idioms in expressions, as shown in the following examples:

* Default Value

image::dataweave-extension-defaultvalue.png[Default Value Code Inspection sample]

* TypeOf

> image::dataweave-extension-typeof.png[TypeOf Code Inspection sample]

Reviewer 4: Can we remove the input directive here?

* IsEmpty

> image::dataweave-extension-isempty.png[Is Empty Code Inspection sample]

Reviewer 4: Can we remove the input directive here?

//image::animations/new_mapping.gif[Creating new Module using DataWeave Extension Plugin]
// ![](animations/new_mapping.gif)

== Test a DataWeave Module

There are two ways to test your module:

* By running a unit test
* By creating a test mapping located in your `src/test/dw` folder to avoid exposing the test to to consumers of your module

The tests use functions of the modules `dw::test::Tests` and `dw::test::Asserts` provided by the artifact `org.mule.weave:data-weave-testing-framework` of the xref:dataweave-testingframework.adoc[DataWeave Testing Framework].

To generate a unit test for a DataWeave module, follow these steps:

. In Visual Studio Code, open your DataWeave project.
. Navigate to the `src/main/dw` folder.
. Select your module `.dwl` file.
. In your module `.dwl` file, click the *Add Unit Test* button located above your function.
+
image::dataweave-extension-testmodule-1.png[MyModule.dwl file showing the function and the Add Unit Test button]

//image::animations/moduleTest.gif[Test Module Steps]
// ![](animations/moduleTest.gif)

This behavior generates:

> * A new `.dwl` test file for the module under the `src/test/dw` folder. This file is where test cases and test suites are described:

Reviewer 2: @kevintroller simplified a bit

** A test case is a single test with the evaluation and assertion of the result.
** A test suite groups tests with a name that better describes the intention of some related cases.
+
The test file defines a test case for the function that you call with specific arguments and use to make assertions over the output.

> * Either a new test suite for the function, consisting of your new test case, or a new test case to an existing test suite.

Reviewer 2: ```suggestion
* A new test case within a test suite for the function. If not present, the test suite is also created. 
```

= Installing and Using the DataWeave 2.0 Extension

Install and use DataWeave 2.0 extension to Visual Studio Code (DataWeave extension) to develop, preview, test, and deploy custom mappings and modules into libraries in a DataWeave project. The extension also enables you to preview transformations similar to those you perform in the https://developer.mulesoft.com/learn/dataweave/[DataWeave Playground].

== Before You Begin

Before installing the DataWeave 2.0 extension to Visual Studio Code, you must:

* Install AdoptOpenJDK (Java 8 minimum)

> * Install Apache Maven (3.6.3 minimum)

Author: I updated these based on your comments, let me know if the minimum versions are correct @martincousido.
+ @dukesphere for visibility on her doc.

Reviewer 2: @kevintroller Looks like this PR says a requirement is "Java 8 or 11 provided," but not sure what that is supposed to mean. https://github.com/mulesoft/mule-dx-data-weave-component/pull/168/files Would a later version of Java work? And I actually installed Oracle Java 11, and it seems to work.

Author: Hi @martincousido  @TomoDC Could you reconfirm the correct requirements to install before installing the extension? It is:

- AdoptOpenJDK (which minimum version) or JDK (Java 8 or 11)

Thank you

. In the *EXTENSIONS* tab, type `dataweave` to search the extension in the marketplace.
. Select *DataWeave 2.0* from the search results.
. Click *Install*.
+
image::dataweave-extension-plugin-install.png[DataWeave 2.0 extension installation]

[start=5]
. After the installation finishes, restart Visual Studio Code.

//image::animations/install_plugin.gif[Installation Steps]
// ![Installation Steps](animations/install_plugin.gif)

== Developing DataWeave Libraries

A DataWeave library is a reusable package of DataWeave module and mapping files, and resource files, such as JSON, XML, and CSV. To develop DataWeave libraries, first create a new DataWeave project in Visual Studio Code where you then create, preview, and test custom mappings and modules.

If you want to try out your mapping without creating a DataWeave project, you can either open your `.dwl` mapping file directly in Visual Studio Code or create a new one and use <<sample-data,sample data>> and <<scenarios,scenarios>>.

=== Understanding the DataWeave Project Structure

DataWeave projects use Apache Maven to build and declare project dependencies. The following diagram shows the DataWeave project structure:

> ```
> ├── src
> │   ├── main
> │   │   ├── dw
> │   │   │   ├── MyMapping.dwl  //Put your DataWeave mapping file here
> |   |   |   └── MyModule.dwl   //Put your DataWeave module file here
> │   │   └── resources
> │   └── test
> │       ├── dw
> │       │   └── Test.dwl  //Unit tests
> │       │   └── MyMapping
> │       │       └── default
> │       │           └── inputs
> │       │               └── payload.json //[Sample data](#sample-data) for your mappings
> │       ├── resources
> ├── pom.xml
> ```

Reviewer 4: ```suggestion
├── src
│   ├── main
│   │   ├── dw
│   │   │   ├── MyMapping.dwl  //Put your DataWeave mappings files here
|   |   |   └── MyModule.dwl   //Put your DataWeave modules files here
│   │   └── resources
│   └── test
│       ├── dw
│       │   └── MyModuleTest.dwl  //Unit tests for the functions in your module
│       │   └── MyMappingTest.dwl //Integration test for your mapping using the saved scenarios
│       ├── resources
│       │       └── MyMapping //Contains the different scenarios for your mapping
│       │           ├── default
│       │           │   └── inputs
│       │           │       └── payload.json //[Sample data](#sample-data) for your mappings
│       │           └── SavedTestScenario //Scenario used for the integration test
├── pom.xml
```
The scenarios are inside test/resources. Added some comments but I'm not sure if this confuses the reader with the concept of scenarios that at this point it is not yet explained.

The tests use functions of the modules `dw::test::Tests` and `dw::test::Asserts` provided by the artifact `org.mule.weave.wtf` of the
//xref:dataweave-testingframework.adoc[DataWeave Testing Framework].

To generate a unit test for a DataWeave module, follow these steps:

. In Visual Studio Code, open your DataWeave project.
. Navigate to the `src/main/dw` folder.
. Select your module `.dwl` file.
. In your module `.dwl` file, click the *Add Unit Test* button located above your function.
+
image::dataweave-extension-testmodule-1.png[MyModule.dwl file showing the function and the Add Unit Test button]

//image::animations/moduleTest.gif[Test Module Steps]
// ![](animations/moduleTest.gif)

This behavior generates:

* A new `.dwl` test file for the module under the `src/test/dw` folder. This file is where test cases and test suites are described:
** A test case is a single test with the evaluation and assertion of the result.
** A test suite groups tests with a name that better describes the intention of some related cases.
+
The test file defines a test case for the function that you call with specific arguments and use to make assertions over the output.

> * Either a new test suite for the function, consisting of your new test case, or a new test case to an existing test suite.

Author: Could you re-review this section @dukesphere @martincousido when you have a moment? Seem like a tongue twister...

Reviewer 2: @kevintroller done, feel free to tweak if you need to

image::dataweave-extension-dependencies.png[Dependencies loading and indexing]

== Create a DataWeave Module

DataWeave modules are `.dwl` files that define functions, variables, types, and namespaces. These files cannot contain output directives, body expressions, or `---` separators between header and body sections. A DataWeave module enables you to share all the content defined in the `.dwl` file for reuse.

To create a new DataWeave module:

. In Visual Studio Code, click *View > Command Palette*.
. In the search bar, type `dataweave`.
. Select *DataWeave: Create New Module*.
+
image::dataweave-extension-createmodule.png[Command Palette with DataWeave: Create new module option selected]
[start=4]
. Add a name for the new module `dwl` file and press enter. +
This creates a new module `.dwl` file under the `src/main/dw` folder.

== Preview a DataWeave Module

> For a better development experience, you can use the _preview_ view to develop your module's functions and mappings. The _preview_ view lets you get instant feedback on how your mapping is working by previewing its output. Use this view to develop the module's functions by creating a DataWeave mapping inside the `src/test/dw` folder that imports your function and calls it to check how is currently working.
> 
> To preview the module functionality, follow these steps:
> 
> . In Visual Studio Code, open your DataWeave project.
> . Navigate to the `src/test/dw` folder.
> . Create a new preview `.dwl` file, for example `MyModulePreview.dwl`. +
>  This new preview file imports the function you created in the `.dwl` file under the `src/main/dw` folder.
> . In the preview `.dwl` file, add your preview mapping code to try out the function.
> . Either preview the module function by one of these options:

Reviewer 4: ```suggestion
For a better development experience, you can use the _preview_ view to develop your module's functions. The _preview_ view lets you get instant feedback on how a mapping file is working by previewing its output. Use this view to develop the module's functions by creating a DataWeave mapping inside the `src/test/dw` folder that imports your functions and calls them to check how are currently working.

To preview the module functionality, follow these steps:

. In Visual Studio Code, open your DataWeave project.
. Navigate to the `src/main/dw` folder.
. Select your module `.dwl` file.
. In your module `.dwl` file, click the *Create Integration Mapping* button located at the top of your module `.dwl` file.
This command is going to create a mapping under the test source folder to test your module, which is not meant to be shared or reused.
. Either preview the module function by one of these options:
```
Using the Cito's codeLense

│           └── SavedTestScenario //Another scenario used for the integration test
```
=== Using the DataWeave Scenarios View

The *DATAWEAVE SCENARIOS* view shows the available scenarios for the `.dwl` file. In this view, you can add, remove, or set the active scenario that is used for your preview execution, or you can choose that the values be autocompleted.

image::dataweave-extension-dataweavescenario.png[DataWeave Scenarios view in Visual Studio Code]

=== Using Reader Properties for Inputs

If you want to set different reader properties to the inputs you use, create a properties file with the suffix `{fileName}-config.properties` in the same directory of your input. For example, if your input is called `payload.csv` then create a `payload-config.properties` file that contains all the reader properties required to run your scenario. Creating the properties file also applies while testing your DataWeave mappings.

== Preview a DataWeave Mapping

After you define sample data, get feedback on how your mapping is working by previewing the output.

. In Visual Studio Code, open your DataWeave project.
. Navigate to the `src/test/dw` folder.

> . Create a new preview `.dwl` file, for example `MyMappingPreview.dwl`. +
>  This new preview file imports the function you created in the `.dwl` file under the `src/main/dw` folder.
> . In the preview `.dwl` file, add your preview mapping code to try out the function.
> . Either preview the mapping function by one of these options:
> * Click the *DataWeave: Run Preview* button (arrow icon) at the top right corner of your preview mapping `.dwl` file.
> 

Reviewer 4: ```suggestion
. Select your mapping `.dwl` file.
. Either preview the mapping function by one of these options:

```
Those steps were just for Previewing module's functions.
Removed duplicate step of "Click the *DataWeave: Run Preview*"
